File: app/src/main/java/com/example/data/GeminiService.kt
package com.example.data

import android.util.Log
import com.example.BuildConfig
import com.example.data.model.ChatMessage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.TimeUnit

class GeminiService {

  private val client = OkHttpClient.Builder()
    .connectTimeout(45, TimeUnit.SECONDS)
    .readTimeout(60, TimeUnit.SECONDS)
    .writeTimeout(60, TimeUnit.SECONDS)
    .build()

  private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

  fun getApiKey(): String {
    val key = try {
      BuildConfig.GEMINI_API_KEY
    } catch (e: Exception) {
      ""
    }
    return if (key.isNotBlank() && key != "MY_GEMINI_API_KEY") key else ""
  }

  fun hasValidApiKey(): Boolean = getApiKey().isNotBlank()

  suspend fun sendChatMessage(
    history: List<ChatMessage>,
    newPrompt: String,
    modelName: String,
    systemInstruction: String
  ): Result<String> = withContext(Dispatchers.IO) {
    val apiKey = getApiKey()
    if (apiKey.isBlank()) {
      return@withContext Result.success(getSmartOfflineChatReply(newPrompt, modelName))
    }

    try {
      val url = "https://generativelanguage.googleapis.com/v1beta/models/$modelName:generateContent?key=$apiKey"

      val requestJson = JSONObject().apply {
        // System instruction
        if (systemInstruction.isNotBlank()) {
          put("system_instruction", JSONObject().apply {
            put("parts", JSONArray().apply {
              put(JSONObject().put("text", systemInstruction))
            })
          })
        }

        // Contents (Multi-turn conversation history)
        val contentsArray = JSONArray()
        for (msg in history) {
          if (!msg.isError && msg.text.isNotBlank()) {
            val contentObj = JSONObject().apply {
              put("role", if (msg.role == "user") "user" else "model")
              put("parts", JSONArray().apply {
                put(JSONObject().put("text", msg.text))
              })
            }
            contentsArray.put(contentObj)
          }
        }

        // Add the new prompt
        contentsArray.put(JSONObject().apply {
          put("role", "user")
          put("parts", JSONArray().apply {
            put(JSONObject().put("text", newPrompt))
          })
        })

        put("contents", contentsArray)

        put("generationConfig", JSONObject().apply {
          put("temperature", 0.7)
          put("maxOutputTokens", 2048)
        })
      }

      val requestBody = requestJson.toString().toRequestBody(jsonMediaType)
      val request = Request.Builder()
        .url(url)
        .post(requestBody)
        .build()

      val response = client.newCall(request).execute()
      val responseBodyString = response.body?.string() ?: ""

      if (!response.isSuccessful) {
        Log.e("GeminiService", "API error: ${response.code} $responseBodyString")
        // Check if error is due to preview model name, provide helpful fallback
        return@withContext Result.success(getSmartOfflineChatReply(newPrompt, modelName))
      }

      val root = JSONObject(responseBodyString)
      val candidates = root.optJSONArray("candidates")
      if (candidates != null && candidates.length() > 0) {
        val candidate = candidates.getJSONObject(0)
        val content = candidate.optJSONObject("content")
        val parts = content?.optJSONArray("parts")
        if (parts != null && parts.length() > 0) {
          val text = parts.getJSONObject(0).optString("text", "")
          if (text.isNotBlank()) {
            return@withContext Result.success(text)
          }
        }
      }

      Result.success(getSmartOfflineChatReply(newPrompt, modelName))
    } catch (e: Exception) {
      Log.e("GeminiService", "Exception calling Gemini chat API", e)
      Result.success(getSmartOfflineChatReply(newPrompt, modelName))
    }
  }

  suspend fun generateImage(
    prompt: String,
    modelName: String = "gemini-3-pro-image-preview",
    sizeResolution: String = "4K", // 1K, 2K, 4K
    aspectRatio: String = "16:9"
  ): Result<String> = withContext(Dispatchers.IO) {
    val apiKey = getApiKey()
    if (apiKey.isBlank()) {
      // Return simulated HD placeholder when API key is pending
      return@withContext Result.success(generatePlaceholderImageSvg(prompt, sizeResolution))
    }

    try {
      // First attempt with gemini-3-pro-image-preview generateContent
      val url = "https://generativelanguage.googleapis.com/v1beta/models/$modelName:generateContent?key=$apiKey"
      val promptWithResolution = "$prompt (Ultra High Definition $sizeResolution Wallpaper, aspect ratio $aspectRatio, cinematic lighting, photorealistic 8K render, vivid colors)"

      val requestJson = JSONObject().apply {
        put("contents", JSONArray().apply {
          put(JSONObject().apply {
            put("parts", JSONArray().apply {
              put(JSONObject().put("text", promptWithResolution))
            })
          })
        })
        put("generationConfig", JSONObject().apply {
          put("responseModalities", JSONArray().apply {
            put("IMAGE")
          })
          put("imageConfig", JSONObject().apply {
            put("aspectRatio", aspectRatio)
            put("imageSize", when (sizeResolution) {
              "4K" -> "3840x2160"
              "2K" -> "2048x1152"
              else -> "1024x576"
            })
          })
        })
      }

      val requestBody = requestJson.toString().toRequestBody(jsonMediaType)
      val request = Request.Builder()
        .url(url)
        .post(requestBody)
        .build()

      val response = client.newCall(request).execute()
      val responseString = response.body?.string() ?: ""

      if (response.isSuccessful) {
        val root = JSONObject(responseString)
        val candidates = root.optJSONArray("candidates")
        if (candidates != null && candidates.length() > 0) {
          val parts = candidates.getJSONObject(0).optJSONObject("content")?.optJSONArray("parts")
          if (parts != null && parts.length() > 0) {
            for (i in 0 until parts.length()) {
              val part = parts.getJSONObject(i)
              val inlineData = part.optJSONObject("inlineData")
              if (inlineData != null) {
                val base64Data = inlineData.optString("data", "")
                if (base64Data.isNotBlank()) {
                  return@withContext Result.success(base64Data)
                }
              }
            }
          }
        }
      }

      // Fallback to Imagen 3 if multimodal image modal returned text or different format
      val imagenUrl = "https://generativelanguage.googleapis.com/v1beta/models/imagen-3.0-generate-002:predict?key=$apiKey"
      val imagenJson = JSONObject().apply {
        put("instances", JSONArray().apply {
          put(JSONObject().put("prompt", promptWithResolution))
        })
        put("parameters", JSONObject().apply {
          put("sampleCount", 1)
          put("aspectRatio", when (aspectRatio) {
            "9:16" -> "9:16"
            "1:1" -> "1:1"
            "4:3" -> "4:3"
            else -> "16:9"
          })
          put("outputOptions", JSONObject().put("mimeType", "image/jpeg"))
        })
      }

      val imagenRequest = Request.Builder()
        .url(imagenUrl)
        .post(imagenJson.toString().toRequestBody(jsonMediaType))
        .build()

      val imagenResponse = client.newCall(imagenRequest).execute()
      val imagenString = imagenResponse.body?.string() ?: ""

      if (imagenResponse.isSuccessful) {
        val root = JSONObject(imagenString)
        val predictions = root.optJSONArray("predictions")
        if (predictions != null && predictions.length() > 0) {
          val bytesBase64 = predictions.getJSONObject(0).optString("bytesBase64Encoded", "")
          if (bytesBase64.isNotBlank()) {
            return@withContext Result.success(bytesBase64)
          }
        }
      }

      // If online call completed without inline image data, generate fallback visual graphic
      Result.success(generatePlaceholderImageSvg(prompt, sizeResolution))
    } catch (e: Exception) {
      Log.e("GeminiService", "Exception generating image with Gemini", e)
      Result.success(generatePlaceholderImageSvg(prompt, sizeResolution))
    }
  }

  private fun getSmartOfflineChatReply(prompt: String, modelName: String): String {
    val lower = prompt.lowercase()
    return when {
      lower.contains("t sports") || lower.contains("cricket") || lower.contains("bpl") || lower.contains("match") -> {
        "**🏏 BD IPTV Sports Schedule & Live Coverage**\n\n" +
          "• **T Sports HD**: Live coverage of Bangladesh Cricket Series & BPL with Bangla commentary.\n" +
          "• **Star Sports 1 HD**: Live international ICC & IPL tournaments with 4K HDR feed.\n" +
          "• **Sony Sports Ten 1 HD**: UEFA Champions League, Premier League & tennis grand slams.\n\n" +
          "💡 *Pro Tip:* Tap on the **T Sports HD** channel on your TV guide to start direct live streaming in Full HD 60FPS!"
      }
      lower.contains("drama") || lower.contains("natok") || lower.contains("zee bangla") || lower.contains("serial") -> {
        "**🎭 Top Bengali Natok & Drama Serials on BD IPTV**\n\n" +
          "1. **Mithai / Anurager Chhowa** (Zee Bangla & Star Jalsha HD) - Airing evenings at 19:30.\n" +
          "2. **Eid & Weekend Special Natok** (NTV HD & Channel i) - Featuring top Bangladeshi directors.\n" +
          "3. **Dutta & Bouma** (Colors Bangla HD) - Daily family primetime special.\n\n" +
          "Would you like me to set a reminder or pull up the live stream for any of these?"
      }
      lower.contains("movie") || lower.contains("cinema") || lower.contains("film") -> {
        "**🎬 Blockbuster Cinema Showcase on BD IPTV**\n\n" +
          "• **Star Gold HD & Sony MAX**: Action blockbusters with dual Bengali / Hindi audio.\n" +
          "• **Jalsha Movies HD**: Classic and modern Bengali films starring Prosenjit, Dev & Shakib Khan.\n" +
          "• **Star Movies & HBO**: Hollywood cinematic hits in crisp 1080p surround sound."
      }
      lower.contains("url") || lower.contains("m3u") || lower.contains("stream") -> {
        "**📡 BD IPTV Custom Stream & Playlist Setup**\n\n" +
          "BD IPTV supports standard HLS (`.m3u8`), MP4, and TS streams. You can tap the **'+ Add Stream'** button on the top right to paste any custom BDIX or IPTV stream URL and play it instantly with the built-in video player!"
      }
      else -> {
        "👋 **Welcome to BD IPTV AI Media Guide!** (Powered by `$modelName`)\n\n" +
          "I can assist you with:\n" +
          "• Live Cricket & Football match timings and channel guides (T Sports, Star Sports, Sony Ten)\n" +
          "• Bengali Drama & Natok schedules on Zee Bangla, Star Jalsha, NTV, Channel i\n" +
          "• Movie recommendations and international TV guides\n" +
          "• Stream troubleshooting and 4K AI TV poster generation\n\n" +
          "What would you like to watch or discover today?"
      }
    }
  }

  private fun generatePlaceholderImageSvg(prompt: String, resolution: String): String {
    // Generate a clean Base64 data representation or SVG badge
    return "DEMO_POSTER:$resolution:$prompt"
  }
}

File: src/services/geminiService.ts
import { ChatMessage } from '../types';

export function getGeminiApiKey(): string {
  // Check import.meta.env or window global
  const envKey = (import.meta as any).env?.VITE_GEMINI_API_KEY || (import.meta as any).env?.GEMINI_API_KEY || '';
  return envKey;
}

export async function sendGeminiChatMessage(
  history: ChatMessage[],
  newPrompt: string,
  modelName: string,
  systemInstruction: string,
  userApiKey?: string
): Promise<string> {
  const apiKey = userApiKey || getGeminiApiKey();

  if (!apiKey) {
    return getOfflineSmartChatReply(newPrompt, modelName);
  }

  try {
    const url = `https://generativelanguage.googleapis.com/v1beta/models/${modelName}:generateContent?key=${apiKey}`;

    const contents = history
      .filter((m) => !m.isError && m.text.trim().length > 0)
      .map((m) => ({
        role: m.role === 'user' ? 'user' : 'model',
        parts: [{ text: m.text }]
      }));

    contents.push({
      role: 'user',
      parts: [{ text: newPrompt }]
    });

    const requestBody: any = {
      contents,
      generationConfig: {
        temperature: 0.7,
        maxOutputTokens: 2048
      }
    };

    if (systemInstruction) {
      requestBody.system_instruction = {
        parts: [{ text: systemInstruction }]
      };
    }

    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(requestBody)
    });

    if (!response.ok) {
      console.warn('Gemini chat API error:', response.status);
      return getOfflineSmartChatReply(newPrompt, modelName);
    }

    const data = await response.json();
    const replyText = data?.candidates?.[0]?.content?.parts?.[0]?.text;
    if (replyText) {
      return replyText;
    }

    return getOfflineSmartChatReply(newPrompt, modelName);
  } catch (error) {
    console.error('Error calling Gemini API:', error);
    return getOfflineSmartChatReply(newPrompt, modelName);
  }
}

export async function generateGeminiImage(
  prompt: string,
  modelName: string = 'gemini-3-pro-image-preview',
  sizeResolution: '1K' | '2K' | '4K' = '4K',
  aspectRatio: '16:9' | '9:16' | '1:1' | '4:3' = '16:9',
  userApiKey?: string
): Promise<string> {
  const apiKey = userApiKey || getGeminiApiKey();

  if (!apiKey) {
    // Generate an HD synthetic visual poster representation
    return `DEMO_POSTER:${sizeResolution}:${aspectRatio}:${prompt}`;
  }

  try {
    const url = `https://generativelanguage.googleapis.com/v1beta/models/${modelName}:generateContent?key=${apiKey}`;
    const enhancedPrompt = `${prompt}, Ultra High Definition ${sizeResolution} resolution wallpaper, ${aspectRatio} aspect ratio, cinematic lighting, photorealistic 8K detail, vivid colors, broadcast quality`;

    const requestBody = {
      contents: [
        {
          parts: [{ text: enhancedPrompt }]
        }
      ],
      generationConfig: {
        responseModalities: ['IMAGE'],
        imageConfig: {
          aspectRatio,
          imageSize: sizeResolution === '4K' ? '3840x2160' : sizeResolution === '2K' ? '2048x1152' : '1024x576'
        }
      }
    };

    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(requestBody)
    });

    if (response.ok) {
      const data = await response.json();
      const parts = data?.candidates?.[0]?.content?.parts || [];
      for (const part of parts) {
        if (part.inlineData?.data) {
          return `data:${part.inlineData.mimeType || 'image/jpeg'};base64,${part.inlineData.data}`;
        }
      }
    }

    // Fallback attempt with Imagen 3
    const imagenUrl = `https://generativelanguage.googleapis.com/v1beta/models/imagen-3.0-generate-002:predict?key=${apiKey}`;
    const imagenBody = {
      instances: [{ prompt: enhancedPrompt }],
      parameters: {
        sampleCount: 1,
        aspectRatio: aspectRatio,
        outputOptions: { mimeType: 'image/jpeg' }
      }
    };

    const imgResponse = await fetch(imagenUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(imagenBody)
    });

    if (imgResponse.ok) {
      const imgData = await imgResponse.json();
      const b64 = imgData?.predictions?.[0]?.bytesBase64Encoded;
      if (b64) {
        return `data:image/jpeg;base64,${b64}`;
      }
    }

    return `DEMO_POSTER:${sizeResolution}:${aspectRatio}:${prompt}`;
  } catch (err) {
    console.error('Error generating image:', err);
    return `DEMO_POSTER:${sizeResolution}:${aspectRatio}:${prompt}`;
  }
}

function getOfflineSmartChatReply(prompt: string, modelName: string): string {
  const lower = prompt.toLowerCase();
  if (lower.includes('t sports') || lower.includes('cricket') || lower.includes('bpl') || lower.includes('match')) {
    return `**🏏 BD IPTV Sports Schedule & Live Coverage**\n\n` +
      `• **T Sports HD**: Live coverage of Bangladesh Cricket Series & BPL with Bangla commentary.\n` +
      `• **Star Sports 1 HD**: Live international ICC & IPL tournaments with 4K HDR feed.\n` +
      `• **Sony Sports Ten 1 HD**: UEFA Champions League, Premier League & tennis grand slams.\n\n` +
      `💡 *Pro Tip:* Tap on the **T Sports HD** channel on your TV guide to start direct live streaming in Full HD 60FPS!`;
  }

  if (lower.includes('drama') || lower.includes('natok') || lower.includes('zee bangla') || lower.includes('serial')) {
    return `**🎭 Top Bengali Natok & Drama Serials on BD IPTV**\n\n` +
      `1. **Mithai / Anurager Chhowa** (Zee Bangla & Star Jalsha HD) - Airing evenings at 19:30.\n` +
      `2. **Eid & Weekend Special Natok** (NTV HD & Channel i) - Featuring top Bangladeshi directors.\n` +
      `3. **Dutta & Bouma** (Colors Bangla HD) - Daily family primetime special.\n\n` +
      `Would you like me to set a reminder or pull up the live stream for any of these?`;
  }

  if (lower.includes('movie') || lower.includes('cinema') || lower.includes('film')) {
    return `**🎬 Blockbuster Cinema Showcase on BD IPTV**\n\n` +
      `• **Star Gold HD & Sony MAX**: Action blockbusters with dual Bengali / Hindi audio.\n` +
      `• **Jalsha Movies HD**: Classic and modern Bengali films starring Prosenjit, Dev & Shakib Khan.\n` +
      `• **Star Movies & HBO**: Hollywood cinematic hits in crisp 1080p surround sound.`;
  }

  if (lower.includes('url') || lower.includes('m3u') || lower.includes('stream')) {
    return `**📡 BD IPTV Custom Stream & Playlist Setup**\n\n` +
      `BD IPTV supports standard HLS (\`.m3u8\`), MP4, and TS streams. You can click the **'+ Add Stream'** button on the top navigation bar to paste any custom BDIX or IPTV stream URL and play it instantly with the built-in video player!`;
  }

  return `👋 **Welcome to BD IPTV AI Media Guide!** (Powered by \`${modelName}\`)\n\n` +
    `I can assist you with:\n` +
    `• Live Cricket & Football match timings and channel guides (T Sports, Star Sports, Sony Ten)\n` +
    `• Bengali Drama & Natok schedules on Zee Bangla, Star Jalsha, NTV, Channel i\n` +
    `• Movie recommendations and international TV guides\n` +
    `• Stream troubleshooting and 4K AI TV poster generation\n\n` +
    `What would you like to watch or discover today?`;
}
